Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Learning Objectives

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# Let's continue to explore feature engineering with the Ames Housing data.  Import the dataset and name it "house".

###1) Import the Ames Housing Data here: https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv

In [1]:
#Code here
import pandas as pd
house = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

###2) Print the first five rows of the house dataset.

In [5]:
#Code here
print(house.shape)
house.head()

(1460, 82)


Id  MSSubClass MSZoning  ...  SaleCondition  SalePrice Large_Lot
0   1          60       RL  ...         Normal     208500       0.0
1   2          20       RL  ...         Normal     181500       0.0
2   3          60       RL  ...         Normal     223500       1.0
3   4          70       RL  ...        Abnorml     140000       0.0
4   5          60       RL  ...         Normal     250000       1.0

[5 rows x 82 columns]

###3) Create a new variable called Large_Lot that takes on the value 1 when LotArea is gerater than or equal to 10,000 and the value 0 when LotArea is less than 10,000.

In [4]:
#Code here
house.loc[house['LotArea'] >= 10000, 'Large_Lot'] = 1
house.loc[house['LotArea'] < 10000, 'Large_Lot'] = 0
house[['LotArea','Large_Lot']] 

LotArea  Large_Lot
0        8450        0.0
1        9600        0.0
2       11250        1.0
3        9550        0.0
4       14260        1.0
...       ...        ...
1455     7917        0.0
1456    13175        1.0
1457     9042        0.0
1458     9717        0.0
1459     9937        0.0

[1460 rows x 2 columns]

###4) Create a new variable called Total_Bathrooms that contains the total number of full and half bathrooms in the house.

In [11]:
#Code here
house.columns 
baths = [ 'BsmtFullBath',
         'BsmtHalfBath',
         'FullBath',
         'HalfBath']
         
house[baths].sum(axis=1)
house['Total_Bathrooms'] = house[baths].sum(axis=1)
baths.append('Total_Bathrooms')
house[baths]

BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  Total_Bathrooms
0                1             0         2         1                4
1                0             1         2         0                3
2                1             0         2         1                4
3                1             0         1         0                2
4                1             0         2         1                4
...            ...           ...       ...       ...              ...
1455             0             0         2         1                3
1456             1             0         2         0                3
1457             0             0         2         0                2
1458             1             0         1         0                2
1459             1             0         1         1                3

[1460 rows x 5 columns]

###5) Create a new variable called Summer_Sale that equals 1 if the sale month was in June, July or August ( MoSold = 6, 7 or 8) and 0 if the sale month was in September - May.

In [51]:
#code here
house.columns 
print(house['MoSold'].value_counts())
# print(house['MoSold'].head())
# print(house['MoSold'].tail())
house.loc[house['MoSold'] == (6 or 7 or 8), 'Summer_Sale'] = 1
house.loc[house['MoSold'] != (6 or 7 or 8), 'Summer_Sale'] = 0
house[['MoSold','Summer_Sale']]


6     253
7     234
5     204
4     141
8     122
3     106
10     89
11     79
9      63
12     59
1      58
2      52
Name: MoSold, dtype: int64


MoSold  Summer_Sale
0          2          0.0
1          5          0.0
2          9          0.0
3          2          0.0
4         12          0.0
...      ...          ...
1455       8          0.0
1456       2          0.0
1457       5          0.0
1458       4          0.0
1459       6          1.0

[1460 rows x 2 columns]

#Now we'll revisit the LendingClub data.

Remember to import the data by first unzipping the file:

In [52]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-10-31 16:56:08--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 52.10.247.232, 52.10.168.94, 34.209.169.217
Connecting to resources.lendingclub.com (resources.lendingclub.com)|52.10.247.232|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [        <=>         ]  22.30M  4.94MB/s    in 5.2s    

2020-10-31 16:56:13 (4.25 MB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [23387713]



In [53]:
!unzip LoanStats_2018Q4.csv.zip


Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


In [54]:
# Read in the CSV
import pandas as pd
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1, skipfooter=2, engine='python')

print(df.shape)
df.head()

(128378, 144)


id  member_id  ...  settlement_percentage  settlement_term
0 NaN        NaN  ...                    NaN              NaN
1 NaN        NaN  ...                    NaN              NaN
2 NaN        NaN  ...                    NaN              NaN
3 NaN        NaN  ...                    NaN              NaN
4 NaN        NaN  ...                    NaN              NaN

[5 rows x 144 columns]

#Approaching a problem two different ways.

In the guided project, we learned how to find the earliest credit year using the built-in pandas date-time format.  However, there is often more than one way to come up with the same solution.  







In the following questions we will work through the steps to create a new variables called Earliest_Credit_Year and issue_year where we use the .split() function remove the month report back just the year as a float data type.  Then we'll use those variables to calculate the length of credit history in years.  

If we do everything correctly, we should calculate *about* the same length of longest credit history that we did in class working with the date-time format.  The answers won't exactly match because we are making different assumptions about when in each month and year the loans were taken out, but it will give you a flavor for different ways of approaching the same problem.

###6) Create a simple test case where credit = 'Jun-1979'.  

Use the .split('-') function to separate the month and year parts of the data and print the year.

Hint: Name the results of the .split('-') function "fields".  fields[0] will return the month part of the date and fields[1] will return the year part of the date.

In [56]:
#code here
credit = 'Jun-1979'
credit.split('-')[1]

'1979'

###7) Use your answer in 6) to write a function called credit_yr that takes in the contents of a cell formatted 'month-year' and returns the year.

Run your function using the simple test case in 6).

In [57]:
#Code here
def credit_yr(monthyear):
  return monthyear.split('-')[1]

credit_yr(credit)



'1979'

###8) Use the .apply() function to apply the function to every cell in the earliest_cr_line variable.

In [61]:
#code here
df['Earliest_Credit_Year'] = df['earliest_cr_line'].apply(credit_yr)

###9) Print the top 5 rows of the Earliest_Credit_Year and earliest_cr_line variables to make sure the variables were created correctly.

In [63]:
#code here
df['Earliest_Credit_Year'].head()

0    2007
1    2002
2    1999
3    2011
4    1993
Name: Earliest_Credit_Year, dtype: object

### 10) Use your function to create a new variable called Issue_Year from the issue_d variable.

In [65]:
#code here
df['Issue_Year'] = df['issue_d'].apply(credit_yr)
df[['issue_d','Issue_Year']]

issue_d Issue_Year
0       Dec-2018       2018
1       Dec-2018       2018
2       Dec-2018       2018
3       Dec-2018       2018
4       Dec-2018       2018
...          ...        ...
128373  Oct-2018       2018
128374  Oct-2018       2018
128375  Oct-2018       2018
128376  Oct-2018       2018
128377  Oct-2018       2018

[128378 rows x 2 columns]

###11) Caluculate length of credit history in years and in days.

Hint: If you are getting an error message, make sure to check the data type of the years output by your functions.

In [72]:
#Code here
print(df[['Earliest_Credit_Year','Issue_Year']])
df['Length_Cr_Hist_Years'] = df['Issue_Year'].astype(int) - df['Earliest_Credit_Year'].astype(int)
df['Length_Cr_Hist_Days'] = (df['Issue_Year'].astype(int) - df['Earliest_Credit_Year'].astype(int))*365.25

       Earliest_Credit_Year Issue_Year
0                      2007       2018
1                      2002       2018
2                      1999       2018
3                      2011       2018
4                      1993       2018
...                     ...        ...
128373                 1985       2018
128374                 1982       2018
128375                 2006       2018
128376                 2008       2018
128377                 2006       2018

[128378 rows x 2 columns]


###12) Calculate the maximum credit length and compare it to we got in the guided project.  

Note - it won't be *exactly* the same because both methods are working in different ways, but they will give you a "pretty close" answer.

In [73]:
#code here
print(df['Length_Cr_Hist_Days'].max())
df['Length_Cr_Hist_Years'].max()

24837.0


68

Guided Project calculated 25,171 days; Assignment calculated 24,837 days

Guided Project calculated 68.91 years; Assignment calculated 68 years

#Portfolio Project Assignment

Start looking for interesting datasets!  

Pretty much every major US governmental agency (CDC, Bureau of Labor Statistics, The Census, etc.) maintains a treasure trove of data.

There's also:

Datasets from the University of Florida Department of Statistics: http://users.stat.ufl.edu/~winner/datasets.html

Pew Research: https://www.pewresearch.org/

General Social Survey: https://gss.norc.org/

Ancestry.com Free data collections: https://www.ancestry.com/search/categories/freeindexacom/?clickref=1101lbhdWxys&key=Uhttps%3A%2F%2Fwww.ancestry.com%2Fsearch%2Fcategories%2Ffreeindexacom%2F%3Fclickref%3D1101lbhdWxys&o_lid=01011l4pVw&o_sch=Affiliate+External&o_xid=01011l4pVw

Popular baby names by year since the founding of the SSA: https://www.babynamewizard.com/ 


https://data.world/

https://www.kaggle.com/datasets

https://catalog.data.gov/dataset

https://datasetsearch.research.google.com/

https://archive.ics.uci.edu/ml/index.php

https://github.com/plotly/datasets

https://lionbridge.ai/datasets/20-best-image-datasets-for-computer-vision/

https://www.visualdata.io/

https://pathmind.com/wiki/open-datasets

https://blog.playment.io/top-open-image-dataset/

https://blog.cambridgespark.com/

https://www.analyticsvidhya.com/blog/2018/03/

https://data-flair.training/blogs/machine-learning-datasets/





